--- VirtualDom.js.orig	2019-06-19 15:38:09.684857715 +0200
+++ VirtualDom.js	2019-06-22 14:07:16.326025005 +0200
@@ -244,6 +244,15 @@ var _VirtualDom_style = F2(function(key,
 		__value: value
 	};
 });
+/* TODO: unused */
+var _VirtualDom_class = F2(function(key, value)
+{
+	return {
+		$: 'a__1_CLASS',
+		__key: key,
+		__value: value
+	};
+});
 var _VirtualDom_property = F2(function(key, value)
 {
 	return {
@@ -400,7 +409,12 @@ function _VirtualDom_organizeFacts(factL
 function _VirtualDom_addClass(object, key, newClass)
 {
 	var classes = object[key];
-	object[key] = classes ? classes + ' ' + newClass : newClass;
+  var newClasses = newClass.split(" ");
+  for (var cs of newClasses) {
+    if (cs === "") continue;
+    object['a__1_CLASS'] = object['a__1_CLASS'] || {};
+    object['a__1_CLASS'][cs] = cs;
+  }
 }
 
 
@@ -484,6 +498,9 @@ function _VirtualDom_applyFacts(domNode,
 		key === 'a__1_STYLE'
 			? _VirtualDom_applyStyles(domNode, value)
 			:
+		key === 'a__1_CLASS'
+			? _VirtualDom_applyClasses(domNode, value)
+			:
 		key === 'a__1_EVENT'
 			? _VirtualDom_applyEvents(domNode, eventNode, value)
 			:
@@ -514,6 +531,25 @@ function _VirtualDom_applyStyles(domNode
 
 
 
+// APPLY CLASSES
+
+
+function _VirtualDom_applyClasses(domNode, classes)
+{
+	var domNodeClassList = domNode.classList;
+
+	for (var key in classes)
+	{
+    if (classes[key] === '') {
+      domNodeClassList.remove(key);
+    } else {
+      domNodeClassList.add(key);
+    }
+  }
+}
+
+
+
 // APPLY ATTRS
 
 
@@ -875,7 +911,7 @@ function _VirtualDom_diffFacts(x, y, cat
 	// look for changes and removals
 	for (var xKey in x)
 	{
-		if (xKey === 'a__1_STYLE' || xKey === 'a__1_EVENT' || xKey === 'a__1_ATTR' || xKey === 'a__1_ATTR_NS')
+		if (xKey === 'a__1_STYLE' || xKey === 'a__1_CLASS' || xKey === 'a__1_EVENT' || xKey === 'a__1_ATTR' || xKey === 'a__1_ATTR_NS')
 		{
 			var subDiff = _VirtualDom_diffFacts(x[xKey], y[xKey] || {}, xKey);
 			if (subDiff)
@@ -897,6 +933,9 @@ function _VirtualDom_diffFacts(x, y, cat
 				(category === 'a__1_STYLE')
 					? ''
 					:
+				(category === 'a__1_CLASS')
+					? ''
+					:
 				(category === 'a__1_EVENT' || category === 'a__1_ATTR')
 					? undefined
 					:
